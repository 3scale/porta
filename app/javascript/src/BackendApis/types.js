// @flow

type ProductLink = {
  name: 'Edit' | 'Overview' | 'Analytics' | 'Methods and Metrics' | 'Mapping Rules',
  path: string
}

export type Backend = {
  id: number,
  name: string,
  systemName: string,
  privateEndpoint: string,
  updatedAt: string,
  links: Array<ProductLink>,
  productsCount: number
}
