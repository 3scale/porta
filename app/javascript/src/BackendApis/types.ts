interface ProductLink {
  name: 'Analytics' | 'Edit' | 'Mapping Rules' | 'Methods and Metrics' | 'Overview';
  path: string;
}

export interface Backend {
  id: number;
  name: string;
  systemName: string;
  privateEndpoint: string;
  updatedAt: string;
  links: ProductLink[];
  productsCount: number;
}
