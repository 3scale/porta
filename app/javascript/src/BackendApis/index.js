// @flow

export * from './components/AddBackendUsage'
export * from './components/BackendSelect'
export * from './components/BackendSelectModal'
export * from './components/PathInput'
