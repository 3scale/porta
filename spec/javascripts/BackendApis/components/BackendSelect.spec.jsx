// @flow

import React from 'react'
import { mount } from 'enzyme'

import { BackendSelect } from 'BackendApis'

const onShowAllSpy = jest.fn()
const onSelectSpy = jest.fn()

const newBackendPath = '/backends/new'
const backends = [
  { id: 0, name: 'API A', privateEndpoint: 'a.com' },
  { id: 1, name: 'API B', privateEndpoint: 'b.com' }
]
const defaultProps = {
  newBackendPath,
  backend: null,
  backends,
  onShowAll: onShowAllSpy,
  onSelect: onSelectSpy
}

const mountWrapper = (props) => mount(<BackendSelect {...{ ...defaultProps, ...props }} />)

afterEach(() => {
  jest.resetAllMocks()
})

it('should render itself', () => {
  const wrapper = mountWrapper({ isOpen: true })
  expect(wrapper.exists()).toBe(true)
})

it('should display all items, a title and a sticky button', () => {
  const wrapper = mountWrapper()
  wrapper.find('.pf-c-select__toggle-button').simulate('click')
  expect(wrapper.find('.pf-c-select__menu li').length).toEqual(backends.length + 2)
})

it('should be able to show a modal', () => {
  const wrapper = mountWrapper()

  wrapper.find('.pf-c-select__toggle-button').simulate('click')
  wrapper.find('.pf-c-select__menu li button').last().simulate('click')

  expect(onShowAllSpy).toBeCalledTimes(1)
  expect(onSelectSpy).toBeCalledTimes(0)
})

it('should be able to select and submit an item', () => {
  const targetItem = backends[0]
  const wrapper = mountWrapper()

  wrapper.find('.pf-c-select__toggle-button').simulate('click')
  wrapper.find('.pf-c-select__menu li button').filterWhere(n => n.text().includes(targetItem.name)).simulate('click')

  expect(onSelectSpy).toBeCalledWith(targetItem)
})

it('should be able to submit the selected backend', () => {
  const targetItem = backends[0]
  const wrapper = mountWrapper({ backend: targetItem })

  const input = wrapper.find('[name="backend_api_config[backend_api_id]"]')
  expect(input.exists()).toBe(true)
  expect(input.prop('value')).toBe(targetItem.id)
})

it('should have a button to create a new backend', () => {
  const wrapper = mountWrapper()
  expect(wrapper.find(`a[href="${newBackendPath}"]`))
})
