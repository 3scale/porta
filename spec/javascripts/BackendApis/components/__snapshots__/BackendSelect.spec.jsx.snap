// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have a table with Name, Private Base URL and Last updated 1`] = `
Array [
  <th
    className=""
    data-key={0}
    data-label=""
    onMouseEnter={[Function]}
  />,
  <th
    className=""
    data-key={1}
    data-label="Name"
    onMouseEnter={[Function]}
    scope="col"
  >
    Name
  </th>,
  <th
    className=""
    data-key={2}
    data-label="Private Base URL"
    onMouseEnter={[Function]}
    scope="col"
  >
    Private Base URL
  </th>,
  <th
    className=""
    data-key={3}
    data-label="Last updated"
    onMouseEnter={[Function]}
    scope="col"
  >
    Last updated
  </th>,
]
`;
